[Quant] Enable XNNPACK ops in QNNPACK BackendConfig

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85863

**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

andrewor14 committed:
[Quant] Enforce constraints on QNNPACK BackendConfig

**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- quant_min_lower_bound = -127 for weight
- quant_max_upper_bound = 127 for weight
- scale_min_lower_bound = 2 ** -12 for both activations and weight

These are consistent with the existing settings in
`default_symmetric_qnnpack_qconfig` and its per_channel and QAT
equivalents, which were added in https://github.com/pytorch/pytorch/pull/74396
and https://github.com/pytorch/pytorch/pull/74507 to enable users
to use this backend with faster XNNPACK quantized ops.

**BC-breaking notes:**

The QConfigs returned by `get_default_qconfig("qnnpack")` and
`get_default_qat_qconfig("qnnpack")` are changed to reflect the
new constraints imposed on the backend. These default QConfigs
are still compatible with the BackendConfig returned by
`get_qnnpack_backend_config()`.

However, existing non-default QConfigs that did not impose the
constraints described above will no longer work with the QNNPACK
BackendConfig. The resulting behavior in this case is that the
corresponding patterns will no longer by quantized, and a warning
explaining what the missing constraints are will be logged.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Enforce constraints on QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for weight
- `quant_max_upper_bound` = 127 for weight
- `scale_min_lower_bound` = 2 ** -12 for both activations and weight

These are consistent with the existing settings in
`default_symmetric_qnnpack_qconfig` and its per_channel and QAT
equivalents, which were added in https://github.com/pytorch/pytorch/pull/74396
and https://github.com/pytorch/pytorch/pull/74507 to enable users
to use this backend with faster XNNPACK quantized ops.

**BC-breaking notes:**

The QConfigs returned by `get_default_qconfig("qnnpack")` and
`get_default_qat_qconfig("qnnpack")` are changed to reflect the
new constraints imposed by the backend. These default QConfigs
are still compatible with the BackendConfig returned by
`get_qnnpack_backend_config()`.

However, existing non-default QConfigs that did not impose the
constraints described above will no longer work with the QNNPACK
BackendConfig. The resulting behavior in this case is that the
corresponding patterns will no longer be quantized, and a warning
explaining what the missing constraints are will be logged.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/backend_config/qnnpack.py:
```diff
@@ -11,63 +11,112 @@
+
+qnnpack_weighted_op_qint8_dtype_config = DTypeConfig(
+    input_dtype=act_qint8_with_constraints,
+    output_dtype=act_qint8_with_constraints,
```
These qint8 `DTypeConfig`s are added to work with `default_symmetric_qnnpack_qconfig` and its per_channel and QAT equivalents

jerryzh168 replied:
I'm wondering if we want to add `qscheme` support to this DTypeConfig as well

andrewor14 replied:
I see, you mean require this one to be `per_tensor_symmetric` or `per_channel_symmetric`?

andrewor14 replied:
Would we want to support affine qint8 weight with affine qint8 activations in the future? (right now we have symmetric qint8 weight with affine qint8 activations) @digantdesai

jerryzh168 replied:
> I see, you mean require this one to be per_tensor_symmetric or per_channel_symmetric?

yeah. this should be a separate PR though

andrewor14 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -972,11 +972,16 @@ def _activation_post_process_satisfies_dtype_config_constraints(
+        if isinstance(activation_post_process, FakeQuantize):
+            observer = activation_post_process.activation_post_process
+        else:
+            observer = activation_post_process
```
This is a small bug fix for https://github.com/pytorch/pytorch/pull/85200, which did not handle `FakeQuantize` correctly

andrewor14 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -972,11 +972,16 @@ def _activation_post_process_satisfies_dtype_config_constraints(
+        if isinstance(activation_post_process, FakeQuantize):
+            observer = activation_post_process.activation_post_process
+        else:
+            observer = activation_post_process
```
This is a bug fix for https://github.com/pytorch/pytorch/pull/85200, which did not handle `FakeQuantize` correctly

jerryzh168 replied:
maybe we can put the logic to get observer into a helper function?

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -239,8 +240,8 @@ def get_default_qconfig(backend='fbgemm', version=0):
         elif backend == 'qnnpack':
-            qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=False),
-                              weight=default_weight_observer)
+            qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=False, eps=2 ** -12),
+                              weight=weight_observer_range_neg_127_to_127)
```
let's separate this change to another PR since this is bc-breaking

andrewor14 committed:
Update on "[Quant] Enforce constraints on QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for weight
- `quant_max_upper_bound` = 127 for weight
- `scale_min_lower_bound` = 2 ** -12 for both activations and weight

These are consistent with the existing settings in
`default_symmetric_qnnpack_qconfig` and its per_channel and QAT
equivalents, which were added in https://github.com/pytorch/pytorch/pull/74396
and https://github.com/pytorch/pytorch/pull/74507 to enable users
to use this backend with faster XNNPACK quantized ops.

**BC-breaking notes:**

The QConfigs returned by `get_default_qconfig("qnnpack")` and
`get_default_qat_qconfig("qnnpack")` are changed to reflect the
new constraints imposed by the backend. These default QConfigs
are still compatible with the BackendConfig returned by
`get_qnnpack_backend_config()`.

However, existing non-default QConfigs that did not impose the
constraints described above will no longer work with the QNNPACK
BackendConfig. The resulting behavior in this case is that the
corresponding patterns will no longer be quantized, and a warning
explaining what the missing constraints are will be logged.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Enable XNNPACK ops on the QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for weight
- `quant_max_upper_bound` = 127 for weight
- `scale_min_lower_bound` = 2 ** -12 for both activations and weight

These constraints will enable users to use this backend with faster
XNNPACK quantized ops and are consistent with the existing settings
in `default_symmetric_qnnpack_qconfig` and its per_channel and QAT
variants. For more detail on why these exact values were chosen,
please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be either per_tensor_symmetric
or per_channel_symmetric.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Enable XNNPACK ops in QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Enable XNNPACK ops in QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/backend_config/qnnpack.py:
```diff
@@ -11,17 +11,50 @@
+    bias_dtype=torch.float,
+)
+
+xnnpack_default_op_qint8_dtype_config = DTypeConfig(
+    input_dtype=xnnpack_act_qint8_with_constraints,
+    output_dtype=xnnpack_act_qint8_with_constraints,
+)
```
cc @digantdesai do we have the min_scale constraints on normal ops as well? or just linear and conv?

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -954,6 +954,18 @@ def find_patterns(
+    if isinstance(activation_post_process, ObserverBase):
+        return activation_post_process
+    else:
+        return activation_post_process.activation_post_process  # type: ignore[return-value]
```
nit: assert that this is a FakeQuantizeBase?

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -954,6 +954,18 @@ def find_patterns(
             user.replace_input_with(last_getitem, new_input)
 
+def _get_observer_from_activation_post_process(
+    activation_post_process: Union[ObserverBase, FakeQuantize],
```
FakeQuantize --> FakeQuantizeBase

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -972,11 +984,12 @@ def _activation_post_process_satisfies_dtype_config_constraints(
             debug_string: str) -> bool:
-        app_quant_min = getattr(activation_post_process, "quant_min", None)
-        app_quant_max = getattr(activation_post_process, "quant_max", None)
+        observer = _get_observer_from_activation_post_process(activation_post_process)
```
I guess the type in L984 should also be FakeQuantizeBase as well

jerryzh168 reviewed (approved):
looks good, had some comments inline

jerryzh168 commented on torch/ao/quantization/backend_config/qnnpack.py:
```diff
@@ -79,16 +112,32 @@ def get_qnnpack_backend_config() -> BackendConfig:
-    conv_dtype_configs = [qnnpack_weighted_op_int8_dtype_config]
+    conv_dtype_configs = [
+        xnnpack_weighted_op_qint8_dtype_config,
+        qnnpack_weighted_op_int8_dtype_config,
```
might make sense to rename int8 to something more specific like quint8

vkuzo commented:
Just curious, what's the user story for this that we want?  Is it:
1. qnnpack is user facing, xnnpack may run some things behind the scenes but user shouldn't care
2. both qnnpack and xnnpack is user facing

Is it (1) or (2), or something else?  From previous conversations with @kimishpatel  and @digantdesai  I remember we were going for (1) - does that sound right?

If it is in fact (1), I think it would be good to name the user facing variables in terms of qnnpack and only mention xnnpack in the comments.  I think the way things named now is a bit confusing, something like `xnnpack_act_qint8_with_constraints` implies that `xnnpack` is user facing.  If we want to do (1), how about `qnnpack_act_qint8_with_constraints_compatible_with_xnnpack_ops`?

kimishpatel commented on torch/ao/quantization/backend_config/qnnpack.py:
```diff
@@ -11,17 +11,50 @@
+    scale_min_lower_bound=2 ** -12,
+)
+
+xnnpack_weight_qint8_with_constraints = DTypeWithConstraints(
```
I dont think we should pollute quantization codebase with backend information anymore than what is already there. So lets not introduce xnnpack here. Can we not just be explicit and name this `weight_qint8_with_symmetric_constraints` and if we want to use backend here maybe say `weight_qint8_with_qnnpack_symmetric_constraints`. Ideally I would like to say `weight_qint8_symmetric_min_scale_2_12` but that is mouthful and confusing.

kimishpatel commented:
> 

Agree with @vkuzo. 1 is what we agreed. As much as possible quant codebase, including user facing api, should not expose xnnpack as yet another backend.

andrewor14 commented:
Sounds good. Just wanted to clarify that xnnpack_act_qint8_with_constraints is not actually user facing (not exposed through __all__), but I can rename it and limit the reference to xnnpack to comments.

vkuzo commented:
> weight_qint8_symmetric_min_scale_2_12

IMO this is totally fine, the variable name should describe the contents.

vkuzo commented:
> Just wanted to clarify that xnnpack_act_qint8_with_constraints is not actually user facing (not exposed through __all__), but I can rename it and limit the reference to xnnpack to comments.

If the story is "there is no user facing xnnpack backend", I think we shoulnd't have any Python variables named after xnnpack.  It should be "backend is qnnpack, and sometimes xnnpack ops are used" consistently throughout the codebase, for both users and developers.  I do think that's a bit confusing, but that's the story we agreed on so we should be consistent.

andrewor14 committed:
Update on "[Quant] Enable XNNPACK ops in QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Enable XNNPACK ops in QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant] Enable XNNPACK ops in QNNPACK BackendConfig"


**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 commented:
Thanks everyone for the feedback. I renamed these to `qnnpack_act_qint8_scale_min_2_neg_12` and `qnnpack_weight_qint8_neg_127_to_127_scale_min_2_neg_12`.

andrewor14 commented:
@pytorchbot merge
